src/avfoundation.cs

```diff
@@ -13513,19 +13513,19 @@ interface AVCompositionTrackFormatDescriptionReplacement : NSSecureCoding {
 		CMFormatDescription ReplacementFormatDescription { get; }
 	}
 
-	delegate /* OSStatus */ int AVAudioSourceNodeRenderHandler (bool isSilence, AudioTimeStamp timestamp, uint frameCount, ref AudioBuffers outputData);
+	delegate /* OSStatus */ int AVAudioSourceNodeRenderHandler2 (ref bool isSilence, ref AudioTimeStamp timestamp, uint frameCount, ref AudioBuffers outputData);
```

Comment:
Here, I'd do the following for XAMCORE_4:
```suggestion
#if !XAMCORE_4
	delegate /* OSStatus */ int AVAudioSourceNodeRenderHandler2 (ref bool isSilence, ref AudioTimeStamp timestamp, uint frameCount, ref AudioBuffers outputData);
#else
#6564 delegate /* OSStatus */ int AVAudioSourceNodeRenderHandler (ref bool isSilence, ref AudioTimeStamp timestamp, uint frameCount, ref AudioBuffers outputData);
```

And we have a nice name when we can break the API.

Reply:
A couple of questions:

1. iirc, `XAMCORE_4` is not defined yet, so we use it more for annotation than functionality. Given this, I'm not sure that the code you've recommended below would compile.

2. Please correct as needed: It seems like the AVCompat.cs file contains C# code (not bindings code) for `[Obsolete]` or broken APIs that we want to "remove" from the code w/o creating breaking changes. For the other APIs in that file, it appears that they are completely removed from the bindings file and moved into the compat file. Is using `#ifdef XAMCORE_4` preferable to using the compat file? Or should I be using both - fwict, these are two different approaches to achieve the same result.

Reply:
You are correct, is a #ifdef but the code would compile since we are just defining the new delegate to be called AVAudioSourceNodeRenderHandler2 now, but not in the future, so we have:

* AVAudioSourceNodeRenderHandler2 - Now 
* AVAudioSourceNodeRenderHandler - Once XAMCORE_4 is defined.

The description of the compat files is correct, but in this case we are not create a method for compatibility reasons, but because it was wrong, so I personally would leave the handler definition were it is, with the fixed sample code I gave. In this case, code locality is important, we keep the delegate signature with the API definition.

tldr; We put inside a #ifdef XAMCORE_4 all those API fixes that we would like to do, but we can since it will break the API. One time, in the future, we will be allow to do it, and we will have nicer APIs. I've heard we had several, but I've never seen one, but XAMCORE_2 etc.. exist.

Reply:
I forgot to mention, we now have to keep both versions, AVAudioSourceNodeRenderHandler and AVAudioSourceNodeRenderHandler2

Reply:
The idea is to:

1. Add a new delegate (`AVAudioSourceNodeRenderHandler2`) with the correct signature.
2. To keep backwards compatibility, we have to keep the old and broken delegate, and all the methods and constructors that use it. Instead we have to add new methods and constructors using the new delegate.
3. The general rule is to move binding code to the AVCompat.cs file when the generated code isn't correct anymore. In your particular case, you want to throw an exception in the old constructors - and that's why those now are in the AVCompat.cs file. In the specific case for the broken delegate itself, it doesn't matter (because there's no generated code you'd want to replace with something else).
4. We also want to have the good delegate name in XAMCORE_4_0 (`AVAudioSourceNodeRenderHandler`), which makes the ifdefs a bit uglier:

```csharp
#if XAMCORE_4_0
delegate int AVAudioSourceNodeRenderHandler (<good signature>);
#else
delegate int AVAudioSourceNodeRenderHandler (<bad signature>);
delegate int AVAudioSourceNodeRenderHandler2 (<good signature>);
#endif
```

Reply:
@rolfbjarne @mandel-macaque Thank you! I think I have made all of the changes you suggested, and all of ^ makes sense.

One more question, `In the specific case for the broken delegate itself, it doesn't matter (because there's no generated code you'd want to replace with something else).`

It seems like the delegate needs to be added to AVCompat.cs. The old delegate cannot be found by the exception throwing methods, even though it is still present in AVFoundation.cs. Why isn't the delegate visible to AVCompat.cs?

Reply:
> It seems like the delegate needs to be added to AVCompat.cs. The old delegate cannot be found by the exception throwing methods, even though it is still present in AVFoundation.cs. Why isn't the delegate visible to AVCompat.cs?

I'm not sure, but it's hard to argue with the compiler... the current code compiles and the api diff look correct, so I wouldn't spend too much time on this unless your curious.

Reply:
If it compiles, lets be happy with it.